README
statsrelay
A consistent-hashing relay for statsd metrics

# License
MIT License
Copyright (c) 2014 Uber Technologies, Inc.

libketama
BSD License
Copyright (c) 2007 Last.fm
Copyright (c) 2007-2014 Richard Jones <rj@metabrew.com>

# Build

Dependencies:
- pkg-config
- libev (>= 4.11)
- glib (>= 2.32)

```
apt-get install libev-dev libglib2.0-dev
make clean
make
make install
```

# Use

Example config:
```
# This is a comment
# Each line is a backend server definition
# IP:PORT WEIGHT
127.0.0.1:8125 600
1.2.3.4:8125 300
# Note: Only dotted quad IPv4 addresses are supported
```

```
statsrelay /path/to/statsrelay.conf
```

This process will run in the foreground. If you need to daemonize, use
start-stop-script, daemontools, supervisord, upstart, systemd, or your
preferred service watchdog.

statsrelay binds to all interfaces on port 8125, tcp and udp.

For each line that statsrelay receives in the statsd format
"statname.foo.bar:1|c\n", the key will be hashed to determine which
backend server the stat will be relayed to. If no connection to that
backend is open, the line is queued and a connection attempt is
started. Once a connection is established, all queued metrics are
relayed to the backend and the queue is emptied. If the backend
connection fails, the queue persists in memory and the connection will
be retried after 5 seconds. Any stats received for that backend during
the retry window are added to the queue.

Each backend has it's own send queue. If a send queue reaches 128MB in
size, new incoming stats are dropped until a backend connection is
successful and the queue begins to drain.

All log messages are sent to syslog with the INFO priority.

Upon SIGHUP, the config file will be reloaded and all backend
connections closed. Note that any stats in the send queue at the time
of SIGHUP will be dropped.

If SIGINT or SIGTERM are caught, all connections are killed, send
queues are dropped, and memory freed. statsrelay exits with return
code 0 if all went well.

To retrieve server statistics, connect to TCP port 8125 and send the
string "status" followed by a newline '\n' character. The end of the
status output is denoted by two consecutive newlines "\n\n"

stats example:
```
nc localhost 8125
status
global bytes_recv_udp counter 0
global bytes_recv_tcp counter 41
global total_connections counter 1
global last_reload timestamp 0
global malformed_lines counter 0
backend:127.0.0.2:8127 bytes_queued counter 27
backend:127.0.0.2:8127 bytes_sent counter 27
backend:127.0.0.2:8127 relayed_lines counter 3
backend:127.0.0.2:8127 dropped_lines counter 0

```


debian/changelog
statsrelay (1.6.8-3) unstable; urgency=medium

  * Revert to old init script

 -- Daniel Chen <dchen@cephalopod>  Mon, 30 Mar 2015 11:07:54 -0700

statsrelay (1.6.8-2) unstable; urgency=medium

  * Fix tag-release.sh

 -- Daniel Chen <dchen@cephalopod>  Wed, 25 Mar 2015 13:34:50 -0700

statsrelay (1.6.8-1) unstable; urgency=medium

  * New upstream release

 -- Daniel Chen <dchen@cephalopod>  Wed, 25 Mar 2015 13:15:05 -0700

statsrelay (1.6.6-1) unstable; urgency=medium

  * New upstream release

 -- Daniel Chen <dchen@cephalopod>  Fri, 13 Mar 2015 17:49:30 -0700

statsrelay (1.6.5-1) unstable; urgency=medium

  * New upstream release

 -- Daniel Chen <dchen@cephalopod>  Fri, 13 Mar 2015 17:28:42 -0700

statsrelay (1.6.4-2) unstable; urgency=medium

  * Update to upstart

 -- Daniel Chen <dchen@cephalopod>  Tue, 03 Mar 2015 16:11:06 -0800

statsrelay (1.6.4-1) unstable; urgency=medium

  * New upstream release

 -- Daniel Chen <dchen@cephalopod>  Mon, 02 Mar 2015 16:02:06 -0800

statsrelay (1.6.3-1) unstable; urgency=medium

  * New upstream release

 -- Daniel Chen <dchen@cephalopod>  Tue, 20 Jan 2015 17:35:49 -0800

statsrelay (1.6.2-2) unstable; urgency=medium

  * Add carbonsink.py to package

 -- Daniel Chen <dchen@cephalopod>  Thu, 15 Jan 2015 18:05:03 -0800

statsrelay (1.6.2-1) unstable; urgency=medium

  * New upstream release

 -- Daniel Chen <dchen@cephalopod>  Thu, 15 Jan 2015 18:04:01 -0800

statsrelay (1.6.1-1) unstable; urgency=low

  * New upstream release

 -- Evan Klitzke <uber@evan.dev.uber.com>  Mon, 12 Jan 2015 23:51:07 +0000

statsrelay (1.6.0-1) unstable; urgency=low

  * New upstream release

 -- Evan Klitzke <uber@evan.dev.uber.com>  Fri, 02 Jan 2015 00:23:12 +0000
statsrelay (1.5.9-2) unstable; urgency=medium

  * Restart daemon after package upgrade

 -- Evan Klitzke <evan@uber.com>  Wed, 31 Dec 2014 14:46:32 -0800

statsrelay (1.5.9-1) unstable; urgency=medium

  * New upstream release

 -- Daniel Chen <dchen@cephalopod>  Wed, 31 Dec 2014 13:16:14 -0800

statsrelay (1.5.8-1) unstable; urgency=medium

  * New upstream release

 -- Daniel Chen <dchen@cephalopod>  Tue, 30 Dec 2014 12:41:33 -0800

statsrelay (1.5.5-1) unstable; urgency=medium

  * New upstream release

 -- Daniel Chen <dchen@cephalopod>  Fri, 26 Dec 2014 17:28:58 -0800

statsrelay (1.5.4-1) unstable; urgency=low

  * New upstream release

 -- Jeremy Grosser <jeremy@synack.me>  Fri, 26 Dec 2014 17:19:19 -0800

statsrelay (1.5.3-1) unstable; urgency=low

  * New upstream release

 -- Jeremy Grosser <jeremy@synack.me>  Fri, 26 Dec 2014 17:10:25 -0800

statsrelay (1.5.2-1) unstable; urgency=low

  * New upstream release

 -- Jeremy Grosser <jeremy@synack.me>  Fri, 26 Dec 2014 16:05:07 -0800

statsrelay (1.5.1-1) unstable; urgency=low

  * New upstream release

 -- Jeremy Grosser <jeremy@synack.me>  Fri, 26 Dec 2014 16:02:05 -0800

statsrelay (1.3.3-1) unstable; urgency=low

  * Add -dbg package, ensure clean upstream

 -- Jeremy Grosser <jeremy@synack.me>  Thu, 11 Dec 2014 14:52:00 -0800

statsrelay (1.3.2-2) unstable; urgency=low

  * Upstream change to configure.ac

 -- Daniel Chen <dchen@uber.com>  Thu, 11 Dec 2014 12:13:02 -0800

statsrelay (1.3.2-1) unstable; urgency=low

  * New upstream release

 -- Daniel Chen <dchen@uber.com>  Thu, 11 Dec 2014 11:24:02 -0800

statsrelay (1.3.1-1) unstable; urgency=low

  * New upstream release

 -- Jeremy Grosser <jeremy@synack.me>  Wed, 10 Dec 2014 17:44:29 -0800

statsrelay (1.3.0-1) unstable; urgency=low

  * New upstream release

 -- Jeremy Grosser <jeremy@synack.me>  Wed, 10 Dec 2014 17:22:56 -0800

statsrelay (1.2.0-1) unstable; urgency=low

  * New upstream release

 -- Jeremy Grosser <jeremy@synack.me>  Fri, 05 Dec 2014 00:43:48 -0800

statsrelay (1.1.8-1) unstable; urgency=low

  * New upstream release

 -- Jeremy Grosser <jeremy@synack.me>  Wed, 03 Dec 2014 17:56:27 -0800

statsrelay (1.1.7-1) unstable; urgency=low

  * New upstream release 

 -- Jeremy Grosser <jeremy@synack.me>  Wed, 26 Nov 2014 15:40:57 -0800

statsrelay (1.1.6-1) unstable; urgency=low

  * New upstream release

 -- Jeremy Grosser <jeremy@synack.me>  Thu, 06 Nov 2014 03:07:38 -0800

statsrelay (1.1.5-1) unstable; urgency=low

  * Bugfix release

 -- Jeremy Grosser <jeremy@synack.me>  Mon, 03 Nov 2014 18:24:34 -0800

statsrelay (1.1.4-1) unstable; urgency=low

  * Add dh-autoreconf as a build dependancy
  * Drop version requirement of glib2.0-dev

 -- David Gillies <dave.gillies@gmail.com>  Thu, 11 Sep 2014 11:33:47 +1000

statsrelay (1.1.4) unstable; urgency=low

  * Add a --no-validation flag

 -- Jeremy Grosser <jeremy@synack.me>  Wed, 10 Sep 2014 17:21:15 -0700

statsrelay (1.1.3) unstable; urgency=low

  * Bugfixes

 -- Jeremy Grosser <jeremy@synack.me>  Fri, 15 Aug 2014 14:44:58 -0700

statsrelay (1.1.2) unstable; urgency=low

  * Handle trailing lines with no ending newline in UDP

 -- Jeremy Grosser <jeremy@synack.me>  Wed, 09 Jul 2014 17:23:32 -0700

statsrelay (1.1.1) unstable; urgency=low

  * Support multi-line UDP packets

 -- Jeremy Grosser <jeremy@synack.me>  Wed, 09 Jul 2014 17:10:13 -0700

statsrelay (1.1.0) unstable; urgency=low

  * Build with -O2

 -- Jeremy Grosser <jeremy@synack.me>  Wed, 09 Jul 2014 15:12:53 -0700

statsrelay (1.0.10) unstable; urgency=low

  * More bugfixes 

 -- Jeremy Grosser <jeremy@synack.me>  Wed, 09 Jul 2014 13:18:44 -0700

statsrelay (1.0.9) unstable; urgency=low

  * Fix segfault due to command line options parsing

 -- Jeremy Grosser <jeremy@synack.me>  Wed, 09 Jul 2014 11:35:06 -0700

statsrelay (1.0.8) unstable; urgency=low

  * Merge fix for init script

 -- Jeremy Grosser <jeremy@synack.me>  Tue, 08 Jul 2014 15:11:03 -0700

statsrelay (1.0.7) unstable; urgency=low

  * Add command line options for config

 -- Jeremy Grosser <jeremy@synack.me>  Tue, 08 Jul 2014 14:56:35 -0700

statsrelay (1.0.6) precise; urgency=low

  * More optimizations

 -- Jeremy Grosser <jeremy@synack.me>  Tue, 08 Jul 2014 18:42:00 +0000


ketama.c
/*
* Copyright (c) 2007, Last.fm, All rights reserved.
* Richard Jones <rj@last.fm>
* Christian Muehlhaeuser <chris@last.fm>
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are met:
*     * Redistributions of source code must retain the above copyright
*       notice, this list of conditions and the following disclaimer.
*     * Redistributions in binary form must reproduce the above copyright
*       notice, this list of conditions and the following disclaimer in the
*       documentation and/or other materials provided with the distribution.
*     * Neither the name of the Last.fm Limited nor the
*       names of its contributors may be used to endorse or promote products
*       derived from this software without specific prior written permission.
*
* THIS SOFTWARE IS PROVIDED BY Last.fm ``AS IS'' AND ANY
* EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
* DISCLAIMED. IN NO EVENT SHALL Last.fm BE LIABLE FOR ANY
* DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
* ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include "ketama.h"

#include <openssl/evp.h>

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <time.h>           /* for reading last time modification   */
#include <unistd.h>         /* needed for usleep                    */
#include <math.h>           /* floor & floorf                       */
#include <sys/stat.h>       /* various type definitions             */
#include <sys/shm.h>        /* shared memory functions and structs  */
#include <sys/types.h>        
#ifdef DEBUG
#include <syslog.h>
#endif
#include <stdarg.h>


char k_error[255] = "";

int num_sem_ids = 0;
int num_shm_ids = 0;
int num_data = 0;
int sem_ids_size = 1024;
int shm_ids_size = 1024;
int shm_data_size = 1024;
int *sem_ids = NULL;
int *shm_ids = NULL;
int **shm_data = NULL;

static void
set_error(const char *format, ...) {
    va_list ap;
    va_start(ap, format);
    vsnprintf(k_error, sizeof(k_error), format, ap);
    va_end(ap);
}

static void
init_sem_id_tracker() {
    sem_ids = malloc(sizeof(int)*1024);
}

static void
init_shm_id_tracker() {
    shm_ids = malloc(sizeof(int)*1024);
}

static void
init_shm_data_tracker() {
    shm_data = malloc(sizeof(int*)*1024);
}

static void
track_shm_data(int *data) {
    if (num_data == shm_data_size) {
        void *tmp = realloc(shm_data, sizeof(int*)*(shm_data_size + 1024));
        if (tmp != NULL) {
            shm_data = tmp;
        } else {
            set_error("Cannot realloc shm data tracker");
            exit(1);
        }

        shm_data_size += 1024;
    }

    shm_data[num_data] = data;
    num_data++;
}

static void
track_sem_id(int semid) {
    if (num_sem_ids == sem_ids_size) {
    void *tmp = realloc(sem_ids, sizeof(int)*(sem_ids_size + 1024));
        if (tmp != NULL) {
            sem_ids = tmp;
        } else {
            set_error("Cannot realloc semids");
            exit(1);
        }

        sem_ids_size += 1024;
    }

    sem_ids[num_sem_ids] = semid;
    num_sem_ids++;
}

static void
track_shm_id(int shmid) {
    if (num_shm_ids == shm_ids_size) {
        void *tmp = realloc(shm_ids, sizeof(int)*(shm_ids_size + 1024));
        if (tmp != NULL) {
            shm_ids = tmp;
        } else {
            set_error("Cannot realloc shmids");
            exit(1);
        }

        shm_ids_size += 1024;
    }

    shm_ids[num_shm_ids] = shmid;
    num_shm_ids++;
}

/** \brief Locks the semaphore.
  * \param sem_set_id The semaphore handle that you want to lock. */
static void
ketama_sem_lock( int sem_set_id )
{
    union semun sem_val;
    sem_val.val = 2;
    semctl( sem_set_id, 0, SETVAL, sem_val );
}


/** \brief Unlocks the semaphore.
  * \param sem_set_id The semaphore handle that you want to unlock. */
static void
ketama_sem_unlock( int sem_set_id )
{
    union semun sem_val;
    sem_val.val = 1;
    semctl( sem_set_id, 0, SETVAL, sem_val );
}


/** \brief Initialize a semaphore.
  * \param key Semaphore key to use.
  * \return The freshly allocated semaphore handle. */
static int
ketama_sem_init( key_t key )
{
    if (sem_ids == NULL) {
        init_sem_id_tracker();
    }

    int sem_set_id;

    sem_set_id = semget( key, 1, 0 );
    track_sem_id(sem_set_id);

    if ( sem_set_id == -1 )
    {
        // create a semaphore set with ID SEM_ID
        sem_set_id = semget( key, 1, IPC_CREAT | 0666 );
        track_sem_id(sem_set_id);

        if ( sem_set_id == -1 )
        {
            strcpy( k_error, "Could not open semaphore!" );
            return 0;
        }

        ketama_sem_unlock( sem_set_id );
    }

    return sem_set_id;
}


/* ketama.h does not expose this function */
void
ketama_md5_digest( char* inString, unsigned char md5pword[EVP_MAX_MD_SIZE] )
{
    EVP_MD_CTX c;

    EVP_MD_CTX_init( &c );
    EVP_DigestInit_ex(&c, EVP_md5(), NULL);
    EVP_DigestUpdate( &c, (void *)inString, strlen( inString ) );
    EVP_DigestFinal_ex( &c, md5pword, NULL );
    EVP_MD_CTX_cleanup( &c );
}


/** \brief Retrieve the modification time of a file.
  * \param filename The full path to the file.
  * \return The timestamp of the latest modification to the file. */
static time_t
file_modtime( char* filename )
{
    struct tm* clock;
    struct stat attrib;

    stat( filename, &attrib );
    clock = gmtime( &( attrib.st_mtime ) );

    return mktime( clock );
}


/** \brief Retrieve a serverinfo struct for one a sever definition.
  * \param line The entire server definition in plain-text.
  * \return A serverinfo struct, parsed from the given definition. */
static serverinfo
read_server_line( char* line )
{
    char* delim = "\t ";
    size_t len;
    serverinfo server;
    server.memory = 0;
    memset(server.addr, '\0', 22);

    char* tok = strtok( line, delim );
    if ( ( strlen( tok ) - 1 ) < 23 )
    {
        char* mem = 0;
        char* endptr = 0;

        strncpy( server.addr, tok, strlen( tok ) );
        server.addr[ strlen( tok ) ] = '\0';

        tok = strtok( 0, delim );
        /* We do not check for a NULL return earlier because strtok will
         * always return at least the first token; hence never return NULL.
         */
        if ( tok == 0 )
        {
            strcpy( k_error, "Unable to find delimiter" );
            server.memory = 0;
        }
        else
        {
            len = strlen(tok);
            mem = (char *)malloc(len + 1);
            memcpy( mem, tok, len + 1);

            errno = 0;
            server.memory = strtol( mem, &endptr, 10 );
            if ( errno == ERANGE || endptr == mem )
            {
                strcpy( k_error, "Invalid memory value" );
                server.memory = 0;
            }

            free( mem );
        }
    }

    return server;
}


/** \brief Retrieve all server definitions from a file.
  * \param filename The full path to the file which contains the server definitions.
  * \param count The value of this pointer will be set to the amount of servers which could be parsed.
  * \param memory The value of this pointer will be set to the total amount of allocated memory across all servers.
  * \return A serverinfo array, containing all servers that could be parsed from the given file. */
static serverinfo*
read_server_definitions( char* filename, unsigned int* count, unsigned long* memory )
{
    serverinfo* slist = 0;
    unsigned int lineno = 0;
    unsigned int numservers = 0;
    unsigned long memtotal = 0;

    FILE* fi = fopen( filename, "r" );
    while ( fi && !feof( fi ) )
    {
        char sline[128] = "";

        if (fgets( sline, 127, fi ) == NULL)
            continue;

        lineno++;

        if ( strlen( sline ) < 2 || sline[0] == '#' )
            continue;

        serverinfo server = read_server_line( sline );
        if ( server.memory > 0 && strlen( server.addr ) )
        {
            slist = (serverinfo*)realloc( slist, sizeof( serverinfo ) * ( numservers + 1 ) );
            memcpy( &slist[numservers], &server, sizeof( serverinfo ) );
            numservers++;
            memtotal += server.memory;
        }
        else
        {
            /* This kind of tells the parent code that
             * "there were servers but not really"
             */
            *count = 1;
            free( slist );
            set_error( "%s (line %d in %s)", k_error, lineno, filename );
            return 0;
        }
    }

    if ( !fi )
    {
        set_error( "File %s doesn't exist!", filename );

        *count = 0;
        return 0;
    }
    else
    {
        fclose( fi );
    }

    *count = numservers;
    *memory = memtotal;
    return slist;
}


unsigned int
ketama_hashi( char* inString )
{
    unsigned char digest[EVP_MAX_MD_SIZE];

    ketama_md5_digest( inString, digest );
    return (unsigned int)(( digest[3] << 24 )
                        | ( digest[2] << 16 )
                        | ( digest[1] <<  8 )
                        |   digest[0] );
}


mcs*
ketama_get_server( char* key, ketama_continuum cont )
{
    unsigned int h = ketama_hashi( key );
    int highp = cont->numpoints;
    mcs (*mcsarr)[cont->numpoints] = cont->array;
    int lowp = 0, midp;
    unsigned int midval, midval1;

    // divide and conquer array search to find server with next biggest
    // point after what this key hashes to
    while ( 1 )
    {
        midp = (int)( ( lowp+highp ) / 2 );

        if ( midp == cont->numpoints )
            return &( (*mcsarr)[0] ); // if at the end, roll back to zeroth

        midval = (*mcsarr)[midp].point;
        midval1 = midp == 0 ? 0 : (*mcsarr)[midp-1].point;

        if ( h <= midval && h > midval1 )
            return &( (*mcsarr)[midp] );

        if ( midval < h )
            lowp = midp + 1;
        else
            highp = midp - 1;

        if ( lowp > highp )
            return &( (*mcsarr)[0] );
    }
}


/** \brief Generates the continuum of servers (each server as many points on a circle).
  * \param key Shared memory key for storing the newly created continuum.
  * \param filename Server definition file, which will be parsed to create this continuum.
  * \return 0 on failure, 1 on success. */
static int
ketama_create_continuum( key_t key, char* filename )
{
    if (shm_ids == NULL) {
        init_shm_id_tracker();
    }

    if (shm_data == NULL) {
        init_shm_data_tracker();
    }

    int shmid;
    int* data;  /* Pointer to shmem location */
    unsigned int numservers = 0;
    unsigned long memory;
    serverinfo* slist;

    slist = read_server_definitions( filename, &numservers, &memory );
    /* Check numservers first; if it is zero then there is no error message
     * and we need to set one. */
    if ( numservers < 1 )
    {
        set_error( "No valid server definitions in file %s", filename );
        return 0;
    }
    else if ( slist == 0 )
    {
        /* read_server_definitions must've set error message. */
        return 0;
    }
#ifdef DEBUG
     syslog( LOG_INFO, "Server definitions read: %u servers, total memory: %lu.\n",
        numservers, memory );
#endif

    /* Continuum will hold one mcs for each point on the circle: */
    mcs continuum[ numservers * 160 ];
    unsigned int i, k, cont = 0;

    for( i = 0; i < numservers; i++ )
    {
        float pct = (float)slist[i].memory / (float)memory;
        unsigned int ks = floorf( pct * 40.0 * (float)numservers );
#ifdef DEBUG
        int hpct = floorf( pct * 100.0 );

        syslog( LOG_INFO, "Server no. %d: %s (mem: %lu = %u%% or %d of %d)\n",
            i, slist[i].addr, slist[i].memory, hpct, ks, numservers * 40 );
#endif

        for( k = 0; k < ks; k++ )
        {
            /* 40 hashes, 4 numbers per hash = 160 points per server */
            char ss[30];
            unsigned char digest[16];

            sprintf( ss, "%s-%d", slist[i].addr, k );
            ketama_md5_digest( ss, digest );

            /* Use successive 4-bytes from hash as numbers
             * for the points on the circle: */
            int h;
            for( h = 0; h < 4; h++ )
            {
                continuum[cont].point = ( digest[3+h*4] << 24 )
                                      | ( digest[2+h*4] << 16 )
                                      | ( digest[1+h*4] <<  8 )
                                      |   digest[h*4];

                memcpy( continuum[cont].ip, slist[i].addr, 22 );
                cont++;
            }
        }
    }
    free( slist );

    /* Sorts in ascending order of "point" */
    qsort( (void*) &continuum, cont, sizeof( mcs ), (compfn)ketama_compare );

    /* Add data to shmmem */
    shmid = shmget( key, MC_SHMSIZE, 0644 | IPC_CREAT );
    track_shm_id(shmid);

    data = shmat( shmid, (void *)0, 0 );
    if ( data == (void *)(-1) )
    {
        strcpy( k_error, "Can't open shmmem for writing." );
        return 0;
    }

    time_t modtime = file_modtime( filename );
    int nump = cont;
    memcpy( data, &nump, sizeof( int ) );
    memcpy( data + 1, &modtime, sizeof( time_t ) );
    memcpy( data + 1 + sizeof( void* ), &continuum, sizeof( mcs ) * nump );

    /* We detatch here because we will re-attach in read-only
     * mode to actually use it. */
#ifdef SOLARIS
    if ( shmdt( (char *) data ) == -1 )
#else
    if ( shmdt( data ) == -1 )
#endif
        strcpy( k_error, "Error detatching from shared memory!" );

    return 1;
}


int
ketama_roll( ketama_continuum* contptr, char* filename )
{
    if (shm_ids == NULL) {
        init_shm_id_tracker();
    }

    if (shm_data == NULL) {
        init_shm_data_tracker();
    }   

    strcpy( k_error, "" );

    key_t key;
    int shmid;
    int *data;
    int sem_set_id;

//     setlogmask( LOG_UPTO ( LOG_NOTICE | LOG_ERR | LOG_INFO ) );
//     openlog( "ketama", LOG_CONS | LOG_PID | LOG_NDELAY, LOG_LOCAL1 );

    key = ftok( filename, 'R' );
    if ( key == -1 )
    {
        set_error( "Invalid filename specified: %s", filename );
        return 0;
    }

    *contptr = malloc( sizeof( continuum ) );
    (*contptr)->numpoints = 0;
    (*contptr)->array = 0;
    (*contptr)->modtime = 0;

    sem_set_id = ketama_sem_init( key );

    int sanity = 0;
    while ( semctl( sem_set_id, 0, GETVAL, 0 ) == 2 )
    {
        // wait for the continuum creator to finish, but don't block others
        usleep( 5 );

        // if we are waiting for > 1 second, take drastic action:
        if(++sanity > 1000000)
    {
            usleep( rand()%50000 );
            ketama_sem_unlock( sem_set_id );
        break;
        }
    }

    time_t modtime = file_modtime( filename );
    time_t* fmodtime = 0;
    while ( !fmodtime || modtime != *fmodtime )
    {
        shmid = shmget( key, MC_SHMSIZE, 0 ); // read only attempt.
        track_shm_id(shmid);

        data = shmat( shmid, (void *)0, SHM_RDONLY );

        if ( data == (void *)(-1) || (*contptr)->modtime != 0 )
        {
            ketama_sem_lock( sem_set_id );

//          if ( (*contptr)->modtime == 0 )
//              syslog( LOG_INFO, "Shared memory empty, creating and populating...\n" );
//          else
//              syslog( LOG_INFO, "Server definitions changed, reloading...\n" );

            if ( !ketama_create_continuum( key, filename ) )
            {
//                 strcpy( k_error, "Ketama_create_continuum() failed!" );
                ketama_sem_unlock( sem_set_id );
                return 0;
            }
/*          else
                syslog( LOG_INFO, "ketama_create_continuum() successfully finished.\n" );*/

            shmid = shmget( key, MC_SHMSIZE, 0 ); // read only attempt.
            track_shm_id(shmid);

            data = shmat( shmid, (void *)0, SHM_RDONLY );
            ketama_sem_unlock( sem_set_id );
        }

        if ( data == (void *)(-1) )
        {
            strcpy( k_error, "Failed miserably to get pointer to shmemdata!" );
            return 0;
        }

        track_shm_data(data);

        (*contptr)->numpoints = *data;
        (*contptr)->modtime = ++data;
        (*contptr)->array = data + sizeof( void* );
        fmodtime = (time_t*)( (*contptr)->modtime );
    }

    return 1;
}


void
ketama_smoke( ketama_continuum contptr )
{
    int i;
    if (shm_data != NULL) {
        for (i = 0; i < num_data; i++) {
            shmdt(shm_data[i]);
        }
        free(shm_data);
        shm_data = NULL;
        num_data = 0;
        shm_data_size = 1024;
    }

    if (sem_ids != NULL) {
        for (i = 0; i < num_sem_ids; i++) {
            semctl(sem_ids[i], 0, IPC_RMID, 0);
        }
        free(sem_ids);
        sem_ids = NULL;
        num_sem_ids = 0;
        sem_ids_size = 1024;
    }

    if (shm_ids != NULL) {
        for (i = 0; i < num_shm_ids; i++) {
            shmctl(shm_ids[i], IPC_RMID, 0);
        }
        free(shm_ids);
        shm_ids = NULL;
        num_shm_ids = 0;
        shm_ids_size = 1024;
    }

    free(contptr);
}


void
ketama_print_continuum( ketama_continuum cont )
{
    int a;
    printf( "Numpoints in continuum: %d\n", cont->numpoints );

    if ( cont->array == 0 )
    {
        printf( "Continuum empty\n" );
    }
    else
    {
        mcs (*mcsarr)[cont->numpoints] = cont->array;
        for( a = 0; a < cont->numpoints; a++ )
        {
            printf( "%s (%u)\n", (*mcsarr)[a].ip, (*mcsarr)[a].point );
        }
    }
}


int
ketama_compare( mcs *a, mcs *b )
{
    return ( a->point < b->point ) ?  -1 : ( ( a->point > b->point ) ? 1 : 0 );
}


char*
ketama_error()
{
    return k_error;
}
